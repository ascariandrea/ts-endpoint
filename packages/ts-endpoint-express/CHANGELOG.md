# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [1.1.0-alpha.9](https://github.com/fes300/open-source/packages/ts-endpoint-express/compare/ts-endpoint-express@1.1.0-alpha.8...ts-endpoint-express@1.1.0-alpha.9) (2020-09-07)


### Features

* **fetch:** add response body to error  "meta" field ([923f412](https://github.com/fes300/open-source/packages/ts-endpoint-express/commit/923f412d26a9d1c3a756219a164a3897d1f8fcac))





# [1.1.0-alpha.8](https://github.com/fes300/open-source/packages/ts-endpoint-express/compare/ts-endpoint-express@1.1.0-alpha.7...ts-endpoint-express@1.1.0-alpha.8) (2020-09-07)


### Features

* send errors as response ([e443ef3](https://github.com/fes300/open-source/packages/ts-endpoint-express/commit/e443ef37b36482a974f0cef453e9053359868e9f))





# [1.1.0-alpha.7](https://github.com/fes300/open-source/packages/ts-endpoint-express/compare/ts-endpoint-express@1.1.0-alpha.6...ts-endpoint-express@1.1.0-alpha.7) (2020-09-04)


### Features

* add infer-fetch-result ([9b1eb75](https://github.com/fes300/open-source/packages/ts-endpoint-express/commit/9b1eb75058881fcb95e9e81551864d7697e7701c))





# [1.1.0-alpha.6](https://github.com/fes300/open-source/packages/ts-endpoint-express/compare/ts-endpoint-express@1.1.0-alpha.5...ts-endpoint-express@1.1.0-alpha.6) (2020-09-04)

**Note:** Version bump only for package ts-endpoint-express





# [1.1.0-alpha.5](https://github.com/fes300/open-source/packages/ts-endpoint-express/compare/ts-endpoint-express@1.1.0-alpha.4...ts-endpoint-express@1.1.0-alpha.5) (2020-09-04)


### Bug Fixes

* uncomment type test ([3497da8](https://github.com/fes300/open-source/packages/ts-endpoint-express/commit/3497da889185878fb8b4c68dcc7c690c01f0eb67))





# [1.1.0-alpha.4](https://github.com/fes300/open-source/packages/ts-endpoint-express/compare/ts-endpoint-express@1.1.0-alpha.3...ts-endpoint-express@1.1.0-alpha.4) (2020-09-04)


### Bug Fixes

* **endpoint:** undefined is valid ([183d709](https://github.com/fes300/open-source/packages/ts-endpoint-express/commit/183d709ac757ac5c2d16b5c84e08d48d60a4e884))





# [1.1.0-alpha.3](https://github.com/fes300/open-source/packages/ts-endpoint-express/compare/ts-endpoint-express@1.1.0-alpha.2...ts-endpoint-express@1.1.0-alpha.3) (2020-09-04)

**Note:** Version bump only for package ts-endpoint-express





# [1.1.0-alpha.2](https://github.com/fes300/open-source/packages/ts-endpoint-express/compare/ts-endpoint-express@1.1.0-alpha.1...ts-endpoint-express@1.1.0-alpha.2) (2020-09-04)

**Note:** Version bump only for package ts-endpoint-express





# [1.1.0-alpha.1](https://github.com/fes300/open-source/packages/ts-endpoint-express/compare/ts-endpoint-express@1.1.0-alpha.0...ts-endpoint-express@1.1.0-alpha.1) (2020-09-04)


### Bug Fixes

* **add-endpoint:** run taskeither ([f20efdf](https://github.com/fes300/open-source/packages/ts-endpoint-express/commit/f20efdf81ae7a9b1299e6b8f2afe6b6139dce1f4))





# 1.1.0-alpha.0 (2020-09-02)


### Bug Fixes

* boh ([695534e](https://github.com/fes300/open-source/packages/ts-endpoint-express/commit/695534e009d51ecd746e4abf22a4efbf63ac9ec6))
* comment test not working in CI ([1ee94d9](https://github.com/fes300/open-source/packages/ts-endpoint-express/commit/1ee94d9617864a552165b625995aa54071df23e2))
* no idea what i am doing ([f7a39bd](https://github.com/fes300/open-source/packages/ts-endpoint-express/commit/f7a39bd5eeb005f9958d1d85594562c60082444e))
* remove unecessary dependency ([f379e99](https://github.com/fes300/open-source/packages/ts-endpoint-express/commit/f379e9964f2cd34415028d7826ed8dcdd00d4fa7))


### Features

* initial implementation ([28bf8f5](https://github.com/fes300/open-source/packages/ts-endpoint-express/commit/28bf8f57803bde5bbf5d64f0da50280dd9653689))
