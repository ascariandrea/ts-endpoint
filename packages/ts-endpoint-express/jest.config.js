const base = require('../../jest.config.base.js');

module.exports = {
  ...base,
  name: 'ts-endpoint-express',
  displayName: 'ts-endpoint-express',
};
