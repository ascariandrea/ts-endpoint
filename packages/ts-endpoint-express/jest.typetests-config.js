const base = require('../../jest.typetests-config.base.js');

module.exports = {
  ...base,
  name: 'ts-endpoint-express',
  displayName: 'ts-endpoint-express',
};
