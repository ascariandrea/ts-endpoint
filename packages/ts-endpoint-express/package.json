{
  "name": "ts-endpoint-express",
  "version": "2.0.0-alpha.13",
  "description": "Adds typesafe endpoints to your express app.",
  "files": [
    "lib"
  ],
  "main": "lib/index.js",
  "repository": "https://github.com/fes300/ts-endpoint/tree/master/packages/ts-endpoint-express",
  "bugs": {
    "url": "https://github.com/fes300/ts-endpoint/issues?q=is%3Aopen+is%3Aissue+project%3Afes300%2Fts-endpoint%2F4"
  },
  "author": "Federico Sordillo <federicosordillo@gmail.com>",
  "license": "ISC",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "postversion": "ls lib && cp package.json lib && cp ../../.npmrc . && npm publish lib",
    "build": "rm -rf ./lib && tsc",
    "type-test": "jest --config=jest.typetests-config.js",
    "runtime-test": "jest --config=jest.config.js",
    "test": "yarn runtime-test --passWithNoTests && yarn type-test"
  },
  "peerDependencies": {
    "express": "^4.17.1",
    "fp-ts": "^2.0.0",
    "io-ts": "^2.2.7"
  },
  "devDependencies": {
    "@types/express": "^4.17.7",
    "express": "^4.17.1"
  },
  "dependencies": {
    "io-ts-types": "^0.5.8",
    "monocle-ts": "^2.3.3",
    "newtype-ts": "^0.3.4",
    "ts-endpoint": "^2.0.0-alpha.10",
    "ts-shared": "^2.0.0-alpha.1",
    "typelevel-ts": "^0.4.0"
  }
}
