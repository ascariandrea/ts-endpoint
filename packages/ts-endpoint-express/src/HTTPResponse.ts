export interface HTTPResponse<T> {
  body: T;
  statusCode: number;
  headers?: Record<string, string>;
}

export interface ResponseError {
  statusCode: number;
  message?: string;
  body?: unknown;
}
