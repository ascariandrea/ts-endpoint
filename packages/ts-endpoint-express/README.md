## ts-endpoint-express

### install

```
$> yarn add ts-endpoint-express
```

### peerDependencies

- "fp-ts": "^2.0.0"
- "io-ts": "^2.2.7"
- "express": "^4.17.1"

### description

`ts-endpoint-express` works in tandem with `ts-endpoint` in order to let you define endpoints just once for both client and server.
It is specifically developed to target the express framework.

### simple guide

First define your endpoints with `ts-endpoint`:

```ts
import * as t from 'io-ts';
import { Endpoint } from 'ts-endpoint';

const getCrayons = Endpoint({
  Input: {
    Params: t.strict({ id: t.string }),
  },
  Method: 'GET',
  getPath: ({ id }) => `users/${id}/crayons`,
  Output: t.strict({ crayons: t.array(t.string) }),
});
```

Then define the endpoints of your express server:

```ts
import { AddEndpoint } from 'ts-endpoint-express';
import * as express from 'express';
import * as TA from 'fp-ts/lib/TaskEither';

const router = express.Router();

AddEndpoint(router)(getCrayons, ({ params: { id } }) => {
  const user: TA.TaskEither<any, { crayons: string[] }> = userService.getByID(id);

  return pipe(
    user,
    TA.map(({ crayons }) => ({ body: { crayons }, statusCode: 200 }))
  );
});
```
