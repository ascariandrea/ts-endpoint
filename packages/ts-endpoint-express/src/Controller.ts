import * as t from 'io-ts';
import { TaskEither } from 'fp-ts/lib/TaskEither';
import { HTTPResponse } from './HTTPResponse';

export type Controller<E, P, H, Q, B, R> = (req: {
  params: P;
  headers: H;
  query: Q;
  body: B;
}) => TaskEither<t.Errors | E, HTTPResponse<R>>;
