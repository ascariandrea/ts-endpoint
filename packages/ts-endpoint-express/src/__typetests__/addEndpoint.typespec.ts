import { AddEndpoint } from 'ts-endpoint-express/src';
import { Endpoint, EndpointError } from 'ts-endpoint';
import * as t from 'io-ts';
import * as express from 'express';
import { left, right } from 'fp-ts/Either';
import { IOError } from 'ts-shared/lib/errors';

const getEndpoint = Endpoint({
  Input: {
    Params: { id: t.string },
    Headers: { auth: t.string },
  },
  Method: 'GET',
  getPath: ({ id }) => `users/${id}/crayons`,
  Output: t.type({ crayons: t.array(t.string) }),
});

const postEndpoint = Endpoint({
  Input: {
    Body: t.type({ content: t.string }),
  },
  Method: 'POST',
  getPath: () => `users/crayons`,
  Output: t.type({ crayons: t.array(t.string) }),
});

const postEndpointWithErrors = Endpoint({
  Input: {
    Body: t.type({ content: t.string }),
  },
  Method: 'POST',
  getPath: () => `users/crayons`,
  Output: t.type({ crayons: t.array(t.string) }),
  Errors: [
    EndpointError(404, t.type({ error: t.string })),
    EndpointError(401, t.type({ baz: t.string })),
  ],
});

const router = express.Router();

// @dts-jest:fail:snap should not allow empty calls
AddEndpoint();

// @dts-jest:fail:snap won't compile if output of controller is wrong
AddEndpoint(router)(getEndpoint, ({ headers: { auth }, params: { id } }) => () => {
  console.log(auth, id);
  return Promise.resolve(right({ body: { crayons: [22] }, statusCode: 200 }));
});

// @dts-jest:fail:snap won't compile if trying to access non existent param
AddEndpoint(router)(getEndpoint, ({ headers: { baz }, params: { id } }) => () => {
  console.log(baz, id);
  return Promise.resolve(right({ body: { crayons: ['brown'] }, statusCode: 200 }));
});

// @dts-jest:fail:snap won't compile if trying to access non defined body
AddEndpoint(router)(getEndpoint, ({ headers: { auth }, params: { id }, body: { foo } }) => () => {
  console.log(auth, id, foo);
  return Promise.resolve(right({ body: { crayons: ['brown'] }, statusCode: 200 }));
});

// @dts-jest:pass:snap correct constructions should work
AddEndpoint(router)(getEndpoint, ({ headers: { auth }, params: { id } }) => () => {
  console.log(auth, id);
  return Promise.resolve(right({ body: { crayons: ['brown'] }, statusCode: 200 }));
});

// @dts-jest:pass:snap correct constructions should work
AddEndpoint(router)(postEndpoint, ({ body: { content } }) => () => {
  console.log(content);
  return Promise.resolve(right({ body: { crayons: ['brown'] }, statusCode: 201 }));
});

// @dts-jest:fail:snap you cannot return a badly formatted error
AddEndpoint(router)(postEndpointWithErrors, ({ body: { content } }) => () => {
  console.log(content);
  return Promise.resolve(left({ foo: 'baz' }));
});

AddEndpoint(router)(postEndpointWithErrors, ({ body: { content } }) => () => {
  console.log(content);
  // @dts-jest:fail:snap you cannot return a badly formatted error
  return Promise.resolve(left(new IOError('error', { kind: 'KnownError', error: 'foo' })));
});

AddEndpoint(router)(postEndpointWithErrors, ({ body: { content } }) => () => {
  console.log(content);
  return Promise.resolve(
    // @dts-jest:fail:snap you cannot return a badly formatted error
    left(new IOError('error', { kind: 'KnownError', status: 401, body: { error: 'foo' } }))
  );
});

// @dts-jest:pass:snap you can return a well formatted error
AddEndpoint(router)(postEndpointWithErrors, ({ body: { content } }) => () => {
  console.log(content);
  return Promise.resolve(
    left(new IOError('error', { kind: 'KnownError', status: 404, body: { error: 'foo' } }))
  );
});
