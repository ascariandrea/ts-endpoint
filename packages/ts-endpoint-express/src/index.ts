import { EndpointInstance, Endpoint } from 'ts-endpoint';
import * as t from 'io-ts';
import * as express from 'express';
import { Controller } from './Controller';
import * as E from 'fp-ts/lib/Either';
import { sequenceS } from 'fp-ts/lib/Apply';
import { pipe } from 'fp-ts/lib/pipeable';
import * as TA from 'fp-ts/lib/TaskEither';

const getRouterMatcher = (
  router: express.Router,
  e: EndpointInstance<any>
): express.IRouterMatcher<express.Router> => {
  switch (e.Method) {
    case 'POST':
      return router.post;
    case 'PATCH':
      return router.patch;
    case 'PUT':
      return router.put;
    case 'DELETE':
      return router.delete;
    case 'GET':
    default:
      return router.get;
  }
};

type CodecRecordOutput<R extends { [k: string]: t.Type<any, any, any> }> = {
  [k in keyof R]: t.TypeOf<R[k]>;
};

type OutputOrNever<T> = T extends { [k: string]: t.Type<any, any, any> }
  ? CodecRecordOutput<T>
  : never;

export type AddEndpoint = (
  router: express.Router,
  ...m: express.RequestHandler[]
) => <
  E,
  O extends t.Type<any, any, any>,
  H extends { [k: string]: t.Type<any, any, any> } | undefined = undefined,
  Q extends { [k: string]: t.Type<any, any, any> } | undefined = undefined,
  B extends { [k: string]: t.Type<any, any, any> } | undefined = undefined,
  P extends { [k: string]: t.Type<any, any, any> } | undefined = undefined
>(
  e: EndpointInstance<Endpoint<any, O, H, Q, B, P>>,
  c: Controller<
    E,
    OutputOrNever<P>,
    OutputOrNever<H>,
    OutputOrNever<Q>,
    OutputOrNever<B>,
    t.TypeOf<O>
  >
) => void;

export const AddEndpoint: AddEndpoint = (router, ...m) => (e, controller) => {
  const matcher = getRouterMatcher(router, e);
  const path = e.getStaticPath((param) => `:${param}`);

  matcher.bind(router)(path, ...(m ?? []), async (req, res, next) => {
    const args = sequenceS(E.either)({
      params: (e.Input.Params ?? t.undefined).decode(req.params),
      headers: (e.Input.Headers ?? t.undefined).decode(req.headers),
      query: (e.Input.Query ?? t.undefined).decode(req.query),
      body: (e.Input.Body ?? t.undefined).decode(req.body),
    });

    const taskRunner = pipe(
      TA.fromEither(args),
      TA.chain((args) => controller(args as any)),
      TA.bimap(
        (e) => next(e),
        (httpResponse) => {
          if (httpResponse.headers !== undefined) {
            res.set(httpResponse.headers);
          }

          return res.status(httpResponse.statusCode).send(httpResponse.body);
        }
      )
    );

    return taskRunner();
  });
};
