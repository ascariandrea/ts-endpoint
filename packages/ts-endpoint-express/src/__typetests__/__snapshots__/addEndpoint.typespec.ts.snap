// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correct constructions should work (type) should match snapshot 1`] = `"void"`;

exports[`correct constructions should work (type) should match snapshot 2`] = `"void"`;

exports[`should not allow empty calls (type) should match snapshot 1`] = `"Expected at least 1 arguments, but got 0."`;

exports[`won't compile if output of controller is wrong (type) should match snapshot 1`] = `
"Type '() => Promise<Left<IOError<never>> | Right<{ body: { crayons: number[]; }; statusCode: number; }>>' is not assignable to type 'TaskEither<IOError<never>, HTTPResponse<{ crayons: string[]; }>>'.
  Type 'Promise<Left<IOError<never>> | Right<{ body: { crayons: number[]; }; statusCode: number; }>>' is not assignable to type 'Promise<Either<IOError<never>, HTTPResponse<{ crayons: string[]; }>>>'.
    Type 'Left<IOError<never>> | Right<{ body: { crayons: number[]; }; statusCode: number; }>' is not assignable to type 'Either<IOError<never>, HTTPResponse<{ crayons: string[]; }>>'.
      Type 'Right<{ body: { crayons: number[]; }; statusCode: number; }>' is not assignable to type 'Either<IOError<never>, HTTPResponse<{ crayons: string[]; }>>'.
        Type 'Right<{ body: { crayons: number[]; }; statusCode: number; }>' is not assignable to type 'Right<HTTPResponse<{ crayons: string[]; }>>'.
          Type '{ body: { crayons: number[]; }; statusCode: number; }' is not assignable to type 'HTTPResponse<{ crayons: string[]; }>'.
            The types of 'body.crayons' are incompatible between these types.
              Type 'number[]' is not assignable to type 'string[]'.
                Type 'number' is not assignable to type 'string'."
`;

exports[`won't compile if trying to access non defined body (type) should match snapshot 1`] = `"Property 'foo' does not exist on type 'undefined'."`;

exports[`won't compile if trying to access non existent param (type) should match snapshot 1`] = `"Property 'baz' does not exist on type 'CodecRecordOutput<{ auth: StringC; }>'."`;

exports[`you can return a well formatted error (type) should match snapshot 1`] = `"<E = never, A = never>(e: E) => Either<E, A>"`;

exports[`you cannot return a badly formatted error (type) should match snapshot 1`] = `
"Type '() => Promise<Right<HTTPResponse<{ crayons: string[]; }>> | Left<{ foo: string; }>>' is not assignable to type 'TaskEither<IOError<{ 404: TypeC<{ error: StringC; }>; 401: TypeC<{ baz: StringC; }>; }>, HTTPResponse<{ crayons: string[]; }>>'.
  Type 'Promise<Right<HTTPResponse<{ crayons: string[]; }>> | Left<{ foo: string; }>>' is not assignable to type 'Promise<Either<IOError<{ 404: TypeC<{ error: StringC; }>; 401: TypeC<{ baz: StringC; }>; }>, HTTPResponse<{ crayons: string[]; }>>>'.
    Type 'Right<HTTPResponse<{ crayons: string[]; }>> | Left<{ foo: string; }>' is not assignable to type 'Either<IOError<{ 404: TypeC<{ error: StringC; }>; 401: TypeC<{ baz: StringC; }>; }>, HTTPResponse<{ crayons: string[]; }>>'.
      Type 'Left<{ foo: string; }>' is not assignable to type 'Either<IOError<{ 404: TypeC<{ error: StringC; }>; 401: TypeC<{ baz: StringC; }>; }>, HTTPResponse<{ crayons: string[]; }>>'.
        Type 'Left<{ foo: string; }>' is not assignable to type 'Left<IOError<{ 404: TypeC<{ error: StringC; }>; 401: TypeC<{ baz: StringC; }>; }>>'.
          Type '{ foo: string; }' is missing the following properties from type 'IOError<{ 404: TypeC<{ error: StringC; }>; 401: TypeC<{ baz: StringC; }>; }>': details, status, name, message"
`;

exports[`you cannot return a badly formatted error (type) should match snapshot 2`] = `
"Argument of type '{ kind: \\"KnownError\\"; error: string; }' is not assignable to parameter of type '{ kind: \\"DecodingError\\"; errors: Errors; } | { kind: CommunicationError; meta?: unknown; status: string; } | { kind: \\"KnownError\\"; status: 404; body: { error: string; }; } | { ...; }'.
  Object literal may only specify known properties, and 'error' does not exist in type '{ kind: \\"DecodingError\\"; errors: Errors; } | { kind: CommunicationError; meta?: unknown; status: string; } | { kind: \\"KnownError\\"; status: 404; body: { error: string; }; } | { ...; }'."
`;

exports[`you cannot return a badly formatted error (type) should match snapshot 3`] = `
"Type '{ error: string; }' is not assignable to type '{ baz: string; }'.
  Object literal may only specify known properties, and 'error' does not exist in type '{ baz: string; }'."
`;
